pkg/services/secrets/kvstore/sql.go

```diff
@@ -52,15 +46,15 @@ func (kv *secretsKVStoreSQL) Get(ctx context.Context, orgId int64, namespace str
 			return nil
 		}
 		isFound = true
-		kv.log.Debug("got secret value", "orgId", orgId, "type", typ, "namespace", namespace)
 		return nil
 	})
 
 	if err == nil && isFound {
 		kv.decryptionCache.Lock()
```

Comment:
I know this isn't code you wrote, but it seems strange to me that the locking and unlocking of the decryption cache happens outside of the decryptionCache's accessors. Wondering if it makes sense to move them. If we want to keep it as-is, I believe it is needed in the remote plugin impl as well.

Reply:
Currently the cache needs to be locked before decryption and unlocked after, decryption is the time consuming task that needs to lock the cache, set/get by themselves is just accessing a variable so it shouldn't need to lock the cache, that's why the .Lock() function is being called outside the accessors before we call `kv.secretsService.Decrypt`

Reply:
> I know this isn't code you wrote, but it seems strange to me that the locking and unlocking of the decryption cache happens outside of the decryptionCache's accessors

I agree. In general you probably want to use a `sync.RWMutex` to lock for either reading or either writing. Maybe related to my question in review why not using 
https://github.com/grafana/grafana/blob/main/pkg/infra/localcache/cache.go

Reply:
I actually didn't know about the existence of `RWMutex` or the `localcache/cache.go` I will refactor it to use the existing cache implementation.

Reply:
I tried refectoring the current cache implementation to use the CacheService in localcache, however, I need to be able to Lock the cache before decryption and Unlock it after decryption, the .Lock and .Unlock are being used in pkg/services/secrets/kvstore/sql.go#L53

This cache implementation is inspired by the precious decryption cache that existed in datasources.

Do you think a mixed implementation where internally my cache implementation uses the CacheService in localcache would be good?

Reply:
I think you need 2 different types of caches and tricky to combine them given the uses. 

One short lived cache to wrap any secret store (sql or plugin) to protect against many "get secret" calls happening in a burst of seconds. This was what I was referring to with 
> In addition, now you have changed both remote plugin and SQL kvstore to use the cache, but given they're implementing the same interface you can implement a "cached kvstore" taking the interface as dependency and adding the cache implementation once.
 
One cache specifically for SQL/default/non-plugin store that does decyption to protect against decrypting the same secret over and over again and by that using up all CPU resources available for a Grafana instance.

Let me know if I've misunderstood something